google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java

```diff
@@ -0,0 +1,568 @@
+    bulkWriter.flush().get();
+  }
+
+  @Test
+  public void addsWritesToNewBatchAfterFlush() throws Exception {
+    RequestResponseMap requestResponseMap =
+        new RequestResponseMap() {
+          {
+            put(
+                batchWrite(create(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                successResponse(2));
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc2")),
+                successResponse(2));
+          }
+        };
+    requestResponseMap.initializeStub(batchWriteCapture, firestoreMock);
+
+    ApiFuture<WriteResult> result1 = bulkWriter.create(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.flush();
+    ApiFuture<WriteResult> result2 = bulkWriter.set(doc2, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+    bulkWriter.close().get();
+
+    List<BatchWriteRequest> requests = batchWriteCapture.getAllValues();
+    assertEquals(requestResponseMap.size(), requests.size());
+
+    verifyRequests(requests, requestResponseMap);
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(2, 0), result1.get().getUpdateTime());
+    assertEquals(Timestamp.ofTimeSecondsAndNanos(2, 0), result2.get().getUpdateTime());
+  }
+
+  @Test
+  public void closeResolvesImmediatelyIfNoWrites() throws Exception {
+    bulkWriter.close().get();
+  }
+
+  @Test
+  public void cannotCallMethodsAfterClose() throws Exception {
+    String expected = "BulkWriter has already been closed.";
+    bulkWriter.close().get();
+    try {
+      bulkWriter.set(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+      fail("set() should have failed");
+    } catch (Exception e) {
+      assertEquals(expected, e.getMessage());
+    }
+    try {
+      bulkWriter.create(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+      fail("create() should have failed");
+    } catch (Exception e) {
+      assertEquals(expected, e.getMessage());
+    }
+    try {
+      bulkWriter.update(doc1, LocalFirestoreHelper.SINGLE_FIELD_MAP);
+      fail("update() should have failed");
+    } catch (Exception e) {
+      assertEquals(expected, e.getMessage());
+    }
+    try {
+      bulkWriter.delete(doc1);
+      fail("delete() should have failed");
+    } catch (Exception e) {
+      assertEquals(expected, e.getMessage());
+    }
+    try {
+      bulkWriter.flush();
+      fail("flush() should have failed");
+    } catch (Exception e) {
+      assertEquals(expected, e.getMessage());
+    }
+    try {
+      bulkWriter.close();
+      fail("close() should have failed");
+    } catch (Exception e) {
+      assertEquals(expected, e.getMessage());
+    }
+  }
+
+  @Test
+  public void sendsWritesToSameDocInSeparateBatches() throws Exception {
+    RequestResponseMap requestResponseMap =
+        new RequestResponseMap() {
+          {
+            put(
+                batchWrite(set(LocalFirestoreHelper.SINGLE_FIELD_PROTO, "coll/doc1")),
+                successResponse(0));
```

Comment:
IMHO 1 is a better starting point since it is different from the default timestamp.